pkg/cmd/search/shared/searcher.go

```diff
@@ -0,0 +1,178 @@
+package shared
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/cli/cli/v2/pkg/search"
+)
+
+type searcher struct {
+	client *http.Client
+	host   string
+}
+
+func NewSearcher(client *http.Client, host string) *searcher {
+	return &searcher{
+		client: client,
+		host:   host,
+	}
+}
+
+func (s *searcher) Search(query search.Query) (search.Result, error) {
+	result := search.Result{}
+	path := fmt.Sprintf("https://api.%s/search/%s", s.host, query.Kind)
+	queryString := url.Values{}
+	q := strings.Builder{}
+	quotedKeywords := quoteKeywords(query.Keywords)
+	q.WriteString(strings.Join(quotedKeywords, " "))
+	for k, v := range query.Qualifiers.ListSet() {
+		v = quoteQualifier(v)
+		q.WriteString(fmt.Sprintf(" %s:%s", k, v))
+	}
+	queryString.Set("q", q.String())
+	if query.Order.IsSet() {
+		queryString.Set(query.Order.Key(), query.Order.String())
+	}
+	if query.Sort.IsSet() {
+		queryString.Set(query.Sort.Key(), query.Sort.String())
+	}
+	pages := math.Ceil((float64(query.Limit) / float64(100)))
+	for i := 1; i <= int(pages); i++ {
+		queryString.Set("page", strconv.Itoa(i))
+		remaining := query.Limit - (i * 100)
+		if remaining > 100 {
+			queryString.Set("per_page", "100")
+		} else if remaining <= 0 {
+			queryString.Set("per_page", strconv.Itoa(query.Limit))
+		} else {
+			queryString.Set("per_page", strconv.Itoa(remaining))
```

Comment:
I find the pagination logic tricky to parse and maybe even unnecessary? The API response will contain the full URL to the next page of results in the `Link` header, so constructing that URL is never necessary. Just keep requesting the next page of results until there are enough results to fill `query.Limit`, or until there are no more pages.

Reply:
Agreed. Using the `Link` header is cleaner, will change.